File: tests/Tests.CSharp/TravelerFrameCrossVerifyTests.cs
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using Xunit;
using Zeta.Core.CSharp;

namespace Zeta.Tests.CSharp;

/// <summary>
/// TravelerFrame cross-language agreement — the C# oracle replays the shared seed
/// (<c>src/Core.TypeScript/traveler-frame/golden-vectors.json</c>) and asserts identical transform /
/// dominates / converge results. Passing == agreeing with the F# oracle on the causal vector-clock frame.
/// </summary>
public class TravelerFrameCrossVerifyTests
{
    private static string RepoRoot()
    {
        var dir = new DirectoryInfo(Path.GetDirectoryName(typeof(TravelerFrameCrossVerifyTests).Assembly.Location)!);
        while (dir is not null && !File.Exists(Path.Join(dir.FullName, "Zeta.sln")))
        {
            dir = dir.Parent;
        }

        return dir?.FullName ?? throw new DirectoryNotFoundException("Could not locate repo root (Zeta.sln).");
    }

    private static Dictionary<string, long> Map(JsonElement obj) =>
        obj.EnumerateObject().ToDictionary(p => p.Name, p => p.Value.GetInt64(), System.StringComparer.Ordinal);

    private static void AssertMapEqual(Dictionary<string, long> expected, IReadOnlyDictionary<string, long> actual)
    {
        Assert.Equal(expected.Count, actual.Count);
        foreach (var kv in expected)
        {
            Assert.True(actual.TryGetValue(kv.Key, out var v) && v == kv.Value, $"key {kv.Key}");
        }
    }

    [Fact]
    public void CSharpTravelerFrameAgreesWithSeed()
    {
        var path = Path.Join(RepoRoot(), "src", "Core.TypeScript", "traveler-frame", "golden-vectors.json");
        Assert.True(File.Exists(path), $"seed not found: {path}");
        using var doc = JsonDocument.Parse(File.ReadAllText(path));
        var seed = doc.RootElement;

        foreach (var v in seed.GetProperty("transform").EnumerateArray())
        {
            AssertMapEqual(Map(v.GetProperty("result")), TravelerFrame.Transform(Map(v.GetProperty("a")), Map(v.GetProperty("b"))));
        }

        foreach (var v in seed.GetProperty("dominates").EnumerateArray())
        {
            Assert.Equal(v.GetProperty("result").GetBoolean(), TravelerFrame.Dominates(Map(v.GetProperty("a")), Map(v.GetProperty("b"))));
        }

        foreach (var v in seed.GetProperty("converge").EnumerateArray())
        {
            var frames = v.GetProperty("frames").EnumerateArray().Select(e => (IReadOnlyDictionary<string, long>)Map(e)).ToList();
            var lub = Map(v.GetProperty("lub"));
            AssertMapEqual(lub, TravelerFrame.Converge(frames));
            AssertMapEqual(lub, TravelerFrame.Converge(Enumerable.Reverse(frames))); // order-independent
        }
    }
}

File: tests/Tests.FSharp/TravelerFrame.CrossVerify.Tests.fs
module Zeta.Tests.TravelerFrameCrossVerifyTests

open System.IO
open System.Text.Json
open global.Xunit
open Zeta.Core

module TF = Zeta.Core.TravelerFrame

// ═══════════════════════════════════════════════════════════════════
// TravelerFrame cross-language agreement — the F# oracle replays the shared seed
// (src/Core.TypeScript/traveler-frame/golden-vectors.json) that the C# oracle (and TS/Rust) also verify.
// Both passing == agreement on the causal vector-clock frame (the 4-lang leg).
// ═══════════════════════════════════════════════════════════════════

let private repoRoot () =
    let mutable dir = DirectoryInfo(Path.GetDirectoryName(System.Reflection.Assembly.GetExecutingAssembly().Location))
    while not (isNull dir) && not (File.Exists(Path.Join(dir.FullName, "Zeta.sln"))) do
        dir <- dir.Parent
    if isNull dir then failwith "Could not locate repo root (Zeta.sln)." else dir.FullName

let private toFrame (e: JsonElement) : TF.Frame =
    { TF.Coords = [ for p in e.EnumerateObject() -> p.Name, Versionstamp.ofInt64 (p.Value.GetInt64()) ] |> Map.ofList }

let private frameMap (f: TF.Frame) : Map<string, int64> = f.Coords |> Map.map (fun _ vs -> vs.Version)
let private toMap (e: JsonElement) : Map<string, int64> =
    [ for p in e.EnumerateObject() -> p.Name, p.Value.GetInt64() ] |> Map.ofList

[<Fact>]
let ``F# TravelerFrame agrees with the shared golden seed`` () =
    let path = Path.Join(repoRoot (), "src", "Core.TypeScript", "traveler-frame", "golden-vectors.json")
    Assert.True(File.Exists path, sprintf "seed not found: %s" path)
    use doc = JsonDocument.Parse(File.ReadAllText path)
    let root = doc.RootElement
    let section (name: string) : JsonElement[] = root.GetProperty(name).EnumerateArray() |> Seq.toArray

    for v in section "transform" do
        Assert.Equal<Map<string, int64>>(toMap (v.GetProperty "result"), frameMap (TF.transform (toFrame (v.GetProperty "a")) (toFrame (v.GetProperty "b"))))

    for v in section "dominates" do
        Assert.Equal<bool>((v.GetProperty "result").GetBoolean(), TF.dominates (toFrame (v.GetProperty "a")) (toFrame (v.GetProperty "b")))

    for v in section "converge" do
        let frames = [ for f in (v.GetProperty "frames").EnumerateArray() -> toFrame f ]
        let lub = toMap (v.GetProperty "lub")
        Assert.Equal<Map<string, int64>>(lub, frameMap (TF.commonFrame frames))
        Assert.Equal<Map<string, int64>>(lub, frameMap (TF.commonFrame (List.rev frames)))
